src/storage/sql_table.cpp

```diff
@@ -12,9 +11,226 @@ namespace terrier::storage {
+                      ProjectedRow *const out_buffer, layout_version_t layout_version) const {
+  // get the version of current tuple slot
+  const auto tuple_version = slot.GetBlock()->data_table_->layout_version_;
+
+  TERRIER_ASSERT(tuple_version <= layout_version,
+                 "The iterator should not go to data tables with more recent version than the current transaction.");
+
+  if (tuple_version == layout_version) {
+    // when current version is same as the intended layout version, get the tuple without transformation
+    return tables_.at(tuple_version).data_table_->Select(txn, slot, out_buffer);
+  }
+
+  // the tuple exists in an older version.
+  // TODO(schema-change): handle versions from add and/or drop column only
+  col_id_t ori_header[out_buffer->NumColumns()];
+
+  auto desired_v = tables_.at(layout_version);
+  auto tuple_v = tables_.at(tuple_version);
+  auto missing UNUSED_ATTRIBUTE = AlignHeaderToVersion(out_buffer, tuple_v, desired_v, &ori_header[0]);
+  auto result = tables_.at(tuple_version).data_table_->Select(txn, slot, out_buffer);
+
+  // copy back the original header
+  std::memcpy(out_buffer->ColumnIds(), ori_header, sizeof(col_id_t) * out_buffer->NumColumns());
+
+  // fill in missing columns and default values
+  FillMissingColumns(out_buffer, desired_v);
+  return result;
+}
+
+std::pair<bool, TupleSlot> SqlTable::Update(const common::ManagedPointer<transaction::TransactionContext> txn,
+                                            RedoRecord *const redo,
+                                            layout_version_t layout_version) const {
+  TERRIER_ASSERT(redo->GetTupleSlot() != TupleSlot(nullptr, 0), "TupleSlot was never set in this RedoRecord.");
+  TERRIER_ASSERT(redo == reinterpret_cast<LogRecord *>(txn->redo_buffer_.LastRecord())
+                             ->LogRecord::GetUnderlyingRecordBodyAs<RedoRecord>(),
+                 "This RedoRecord is not the most recent entry in the txn's RedoBuffer. Was StageWrite called "
+                 "immediately before?");
+  // TODO(Schema-Change): need to take care of tuple migration if the update touches the new columns added
+  //  The migration should be a delete (MVCC style) in old datatable followed by an insert in new datatable.
+
+  // get the version of current tuple slot
+  auto curr_tuple = redo->GetTupleSlot();
+  auto returned_slot = curr_tuple;
+
+  const auto tuple_version = curr_tuple.GetBlock()->data_table_->layout_version_;
+
+  TERRIER_ASSERT(tuple_version <= layout_version,
+                 "The iterator should not go to data tables with more recent version than the current transaction.");
+
+  bool result;
+  if (tuple_version == layout_version) {
+    result = tables_.at(layout_version).data_table_->Update(txn, curr_tuple, *(redo->Delta()));
+  } else {
+    // tuple in an older version, check if all modified columns are in the datatable version where the tuple is in
+
+    col_id_t ori_header[redo->Delta()->NumColumns()];
```

Comment:
would suggest renaming this to orig_header for readability

Reply:
ok